### ISSUE TITLE
Fail to create reproducible model

### ISSUE THREAD

**Issue by JoseSanchezVillicana** [scverse/scvi-tools#3611]
Hi scivi-tools team,

I am currently working on an MrVI model using scRNA-Seq data from two different publications. I was following the [Quick Start Tutorial](https://docs.scvi-tools.org/en/latest/tutorials/notebooks/scrna/MrVI_tutorial.html) and I set the random seed just like it is recommended there. But I run my code two times one after the other and I get two different outputs for the local sample distances. These distances do not vary too much, but there is clearly a difference between them. 

I have even attempted to set other possible sources of randomness manually using the same seed, but this results in the same problem. 

I am working for an upcoming publication and reproducibility is really important. I would really appreciate if anyone could tell me whether I am missing something or how to fix the problem.

These are the potential sources of randomness I am attempting to manually control.

```python
# For reproducibility purposes
seed = 42

# Python / NumPy / Random
random.seed(seed)
np.random.seed(seed)
os.environ["PYTHONHASHSEED"] = str(seed)

# PyTorch
torch.manual_seed(seed)
torch.cuda.manual_seed_all(seed)
torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False

# JAX
jax.random.PRNGKey(seed)
os.environ['JAX_ENABLE_X64'] = '0'
os.environ['JAX_PLATFORM_NAME'] = 'gpu'

# scivi-tools global seed
scvi.settings.seed = seed

# Disable JAX async dispatch (prevents nondeterminism)
os.environ['JAX_DISABLE_JIT'] = '0'
os.environ['JAX_TRACEBACK_FILTERING'] = 'off'
```

And these is the code I am using to create and train the model.
```python
# Setup
adata.obs['MrVI_batch'] = (
    adata.obs['Technique'].astype(str)
    + '_' +
    adata.obs['Sample ID'].astype(str)
)

batch_key = 'MrVI_batch'
sample_key = 'subannotation'
labels_key = None

MRVI.setup_anndata(
    adata,
    sample_key= sample_key,
    batch_key= batch_key
)

# Creation and training
model = MRVI(adata, n_latent= 30)
model.train(accelerator= 'gpu')
```

#### Versions:

> VERSION 1.3.3

**Comment by ori-kron-wis** [comment]
When you say you run them one after the other, did you completely clear your session between runs? i.e, restart the notebook and clear all cache and outputs if working on a notebook, and/or close the console and start a Python session again if from CLI or IDE. You need to run the whole code once, start to end, is the only way to be sure if there are reproducible results.

If you are under the same session and running the same code w/o a restart it or cell by cell if on a notebook - the results will not be reproducible, despite those many seeds you made, they will be close though.

**Comment by JoseSanchezVillicana** [comment]
> When you say you run them one after the other, did you completely clear your session between runs? i.e, restart the notebook and clear all cache and outputs if working on a notebook, and/or close the console and start a Python session again if from CLI or IDE. You need to run the whole code once, start to end, is the only way to be sure if there are reproducible results.
> 
> If you are under the same session and running the same code w/o a restart it or cell by cell if on a notebook - the results will not be reproducible, despite those many seeds you made, they will be close though.

Hey, thank you for the quick response! I am working on a notebook, and I restarted the kernel and cleared all outputs between runs. Indeed, the results are close, but for some cell types, the variation is significant.

**Comment by ori-kron-wis** [comment]
@JoseSanchezVillicana I verified that our mrvi tutorial reproduces fully if I rerun it after restarting and clearing the cache. That includes get_local_sample_distances part, there was no need for additional seeding.

What I can suggest is to update to most recent version and run the mrvi torch version (as we stopped supporting the jax version). Plus, verify that the tutorial reproduces as well at your side, like i did.

After that, continue with your personal analysis.

Is there any additional functions you want me to check?

**Comment by JoseSanchezVillicana** [comment]
@ori-kron-wis I have rerun the analysis using the torch backend, but it is still not the same results between them.

Also, I have come up with this message (when running the get_local_samples method) I thought to share: `/home/villicana/.conda/envs/plotting_app/lib/python3.11/site-packages/torch/nn/functional.py:6487: UserWarning: There is a performance drop because we have not yet implemented the batching rule for aten::_scaled_dot_product_flash_attention_for_cpu. Please file us an issue on GitHub so that we can prioritize its implementation. (Triggered internally at /pytorch/aten/src/ATen/functorch/BatchedFallback.cpp:81.)`

**Comment by ori-kron-wis** [comment]
Can you please attach what exactly is different that you see (images of it)?

**Comment by JoseSanchezVillicana** [comment]
I forgot to mention I am also unable to reproduce the tutorial you have.

These are the plots I generate to check if the models are reproduced the same way. I restarted and cleared all outputs between runs. I am now running the v: 1.4.0.post1 (because I updgraded as you recommended). I am running the MrVI with Torch backend. The code I use to generate the plots is exactly the same between runs.

The distances should be the same everytime, right?

Distance dist across cell types (1st run & 2nd run):
<img width="722" height="517" alt="Image" src="https://github.com/user-attachments/assets/17a82217-402c-41c3-b8d9-80934e9e18a5" />

<img width="1048" height="531" alt="Image" src="https://github.com/user-attachments/assets/d1196f35-24e6-4f66-90a0-d4a6864bb3a5" />

Counterfactual distance from all cell types to three specific cell types target (1st run & 2nd run):
<img width="886" height="308" alt="Image" src="https://github.com/user-attachments/assets/58a2a1c2-eb62-402d-99ff-acb5ecbef9fc" />

<img width="812" height="310" alt="Image" src="https://github.com/user-attachments/assets/6440f9ba-9ed2-4d78-b3e2-8bf8a587baaf" />

Mean distance from cell types to the same targets (1st run & 2nd run):
<img width="427" height="814" alt="Image" src="https://github.com/user-attachments/assets/8e9f994c-aa9e-4359-9056-6e19d3a3f5b3" />

<img width="438" height="807" alt="Image" src="https://github.com/user-attachments/assets/b1b5ef15-dc8e-490f-88fd-e080563a29fa" />

**Comment by ori-kron-wis** [comment]
Can you share your code here or privately (at least the downstream plotting)? I want to see what exactly you did and apply it also to see if I get changes as well

**Comment by JoseSanchezVillicana** [comment]
After the training (just the first plot I shared):

```
# Retaining cell to cell distances
dists = model.get_local_sample_distances(
    adata,
    keep_cell= True,
    groupby= 'subannotation',
    normalize_distances= False,
    use_mean= True
)

# Avg distance from other samples per cell: vector to quantify sample effect heterogeneity within cell populations
cell_dists = dists['cell']
dist_vec = []
celltype_anno = []
sample_anno = []
target_anno = []

for i in range(cell_dists.shape[0]):
    subannot_temp = adata.obs['subannotation'].iloc[i]
    subannot_temp_index = np.where(cell_dists['sample_x'].to_numpy() == subannot_temp)[0][0]
    vec_temp = cell_dists[i, subannot_temp_index, :].to_numpy()
    dist_vec.append(vec_temp)
    celltype_anno.append([subannot_temp]*len(vec_temp))
    sample_anno.append([adata.obs['subannotation'].iloc[i]] * len(vec_temp))
    target_anno.append(cell_dists[i,:]['sample_y'].values.tolist())

dist_vec = np.concatenate(dist_vec, axis=0)
celltype_anno = np.concatenate(celltype_anno, axis=0)
sample_anno = np.concatenate(sample_anno, axis=0)
target_anno = np.concatenate(target_anno, axis=0)
# Overall distribution of distances
plt.hist(dist_vec, bins= 100)
plt.show()
```

**Comment by ori-kron-wis** [comment]
> Distance dist across cell types (1st run & 2nd run):

Well, here you have a completely different kind of plots. So what exactly are you comparing? As I said, in a jupyter setting, everything is important for reproducibility. You need to run **exactly** the same thing. Right now, I see different figures, resulting from different methods or scripts (?) so all other things are not reproducible just because of that (and probably there are other things).

Having said that, we don't expect large differences even for different environments and settings, and here we see large differences. Has your model even converged? seem other issues might exist before we even try to take care of the reproducibility thing. For me, 1st plot is like model stopped training before convergence.
I think what we see are issues with model training/data and not reproducibility (which is by our definition, how we reproduce results of same code, given model and data are ok).

> # Retaining cell to cell distances
> dists = model.get_local_sample_distances(
>     adata,
>     keep_cell= True,
>     groupby= 'subannotation',
>     normalize_distances= False,
>     use_mean= True
> )

To clarify: most of the parameters you chose here, are just default. The only thing you needed is groupby= 'subannotation'

> I forgot to mention, I am also unable to reproduce the tutorial you have.

That's the first thing to check. The tutorial is fully reproduced at our hand (make sure you restart the kernel and clear the cache between runs), including that piece of code you added here and its plot. 
I just made 2 copies of the exact same file and ran them side by side in different tabs and compare. We look there are no random effect that affect the results.

> Also, I have come up with this message (when running the get_local_samples method) I thought to share: `/home/villicana/.conda/envs/plotting_app/lib/python3.11/site-packages/torch/nn/functional.py:6487: UserWarning: There is a performance drop because we have not yet implemented the batching rule for aten::_scaled_dot_product_flash_attention_for_cpu. Please file us an issue on GitHub so that we can prioritize its implementation. (Triggered internally at /pytorch/aten/src/ATen/functorch/BatchedFallback.cpp:81.)`

I don't know functorch and I don't know if it might be related to what you see, but we don't use that, so can't help with it. 
I can suggest trying also the JAX version of mrvi and see if you still can't reproduce/get the results you expected.

> Counterfactual distance from all cell types to three specific cell types target (1st run & 2nd run):

You didn't send this code, but it seems the issue is before that.
